### PR TITLE
5 nni hashtables

### PR DESCRIPTION
**Pull request by thayerf**
5 newick reorder is also subsumed in this branch. Cuckoo hash tables store all possible NNI moves and their respective likelihoods for a given topology.

### CODE REVIEW

File: src/partition.cpp [hunk 1]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
-
 /// @brief Returns a the tree as a Newick string.
 /// @return Newick string.
-std::string Partition::ToNewick() {
-  return std::string(pll_utree_export_newick(tree_));
+std::string Partition::ToNewick(pll_utree_t *tree) {
+  return std::string(utree_short_newick(tree));
+}
+
+/// @brief Recursive function for newick string without branch lengths.
```
Comment (matsen) [inline src/partition.cpp]: "for newick string" -> "to generate Newick string" ?

Comment (matsen) [inline src/partition.cpp]: I would suggest (in the longer-form documentation) noting that this actually sprintfs the utree as if it was a rooted tree.

File: src/partition.cpp [hunk 2]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+  if (size_alloced < 0) {
+    pll_errno = PLL_ERROR_MEM_ALLOC;
+    snprintf(pll_errmsg, 200, "memory allocation during newick export failed");
+    return NULL;
+  }
+
+  return newick;
+}
+
+/// @brief Prints newick string without branch lengths.
```
Comment (matsen) [inline src/partition.cpp:126]: ... whereas this one prints as an unrooted tree.

File: src/partition.cpp [hunk 3]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+
+  return (newick);
+}
+
+/// @brief Finds node in tree's subtree with the label that is first
+/// alphabetically.
+/// @param[in] tree
+/// Parent node of subtree to root.
+/// @return smallest label
+std::string Partition::RootNewickRecursive(pll_utree_t *tree) {
```
Comment (matsen) [inline src/partition.cpp]: The function name doesn't seem like a particularly accurate description of what the function does, does it?

File: src/partition.cpp [hunk 4]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+  /*Root the tree at parent of lowest-labeled leaf*/
```
Comment (matsen) [inline src/partition.cpp]: ... nor does this comment

File: src/partition.cpp [hunk 5]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+  if (!tree->next) {
+    if (tree->label == label) {
+      root[0] = tree->back;
+      return true;
+    } else
+      return false;
+  }
+  if (SetLabelRoot(label, tree->next->back, root))
+    return true;
+  if (SetLabelRoot(label, tree->next->next->back, root))
```
Comment (matsen) [inline src/partition.cpp]: Combine this with the previous statement using or?

File: src/partition.cpp [hunk 6]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+
+  return minlabel;
+}
+
+/// @brief Finds a node with the given label and sets the current node to the
+/// parent of it.
+/// @param[in] label
+/// Label of node to find
+/// @param[in] tree
+/// Node to search at
```
Comment (matsen) [inline src/partition.cpp:194]: Missing a param here. You should be able to run Doxygen without errors. (Doesn't this show an error?)

File: src/partition.cpp [hunk 7]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+  }
+  if (SetLabelRoot(label, tree->next->back, root))
+    return true;
+  if (SetLabelRoot(label, tree->next->next->back, root))
+    return true;
+  return false;
+}
+
+/// @brief Sets the label of the root for the entire tree. Sets tree_ to that
+/// node.
```
Comment (matsen) [inline src/partition.cpp]: Does it really?

File: src/partition.cpp [hunk 8]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+/// @param[in] tree
+/// An internal node.
```
Comment (matsen) [inline src/partition.cpp:213]: Needs return descriptor.

File: src/partition.cpp [hunk 9]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
 }
 
+/// @brief Determines order for first ternary step, then runs recursive ordering
+/// for the two non-root subtrees.
+/// @return Completely ordered newick string.
+pll_utree_t *Partition::ToOrderedNewick(pll_utree_t *tree) {
+  tree = SetNewickRoot(tree);
+  RootNewickRecursive(tree);
+
+  /*Check tree health after reordering*/
```
Comment (matsen) [inline src/partition.cpp]: Do we actually expect that there could be problems with health after reordering? If no, perhaps take this out and turn it into a test?

File: src/partition.cpp [hunk 10]
```diff
@@ -128,31 +285,35 @@ void Partition::FullTraversalUpdate(pll_utree_t* tree) {
...
 /// @brief Just calculate log likelihood.
 /// NOTE: This will return nonsense if you haven't updated CLV's, etc.
 /// @return Log likelihood.
-double Partition::LogLikelihood() {
+double Partition::LogLikelihood(pll_utree_t *tree) {
   double logl = pll_compute_edge_loglikelihood(
-      partition_, tree_->clv_index, tree_->scaler_index, tree_->back->clv_index,
-      tree_->back->scaler_index, tree_->pmatrix_index, params_indices_,
-      NULL);  // Can supply a persite_lnl parameter as last argument.
+      partition_, tree->clv_index, tree->scaler_index, tree->back->clv_index,
```
Comment (matsen) [inline src/partition.cpp:295]: Why member variables for some arguments and `tree` for others?

File: src/partition.cpp [hunk 11]
```diff
@@ -201,52 +370,144 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
   return len;
 }
 
-
+/// @brief Aux function for optimizing branch lengths accross the whole tree.
```
Comment (matsen) [inline src/partition.cpp]: spelling

File: src/partition.cpp [hunk 12]
```diff
@@ -201,52 +370,144 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
-  OptimizeCurrentBranch(tree);
-  std::cout<<"DONE"<<std::endl;
-
-  TreeBranchLengthsAux(tree->next->back);
-  TreeBranchLengthsAux(tree->next->next->back);
+  if (!tree->next) {
+
+    TraversalUpdate(tree->back, 1);
+    OptimizeCurrentBranch(tree->back);
+    // std::cout<<"DONE"<<std::endl;
```
Comment (matsen) [inline src/partition.cpp]: Cut

File: src/partition.cpp [hunk 13]
```diff
@@ -201,52 +370,144 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
   }
 }
 
-
-void Partition::TreeBranchLengths(){
-  if(!tree_->next)
-    {
+///@brief Perform a postorder tree traversal, optimizing the branch length at
+/// every edge.
+/// NOTE: This optimizes initial branch twice per traversal.
```
Comment (matsen) [inline src/partition.cpp]: grammar

File: src/partition.cpp [hunk 14]
```diff
@@ -201,52 +370,144 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
-
-  int i=0;
-  while((fabs(loglike_prev-loglike) > EPSILON)&& i<MAX_ITER){
-  TreeBranchLengths();
-
-  loglike_prev=loglike;
-  loglike=FullTraversalLogLikelihood();
-  i++;
+  int i = 0;
+  while ((fabs(loglike_prev - loglike) > EPSILON) && i < MAX_ITER) {
```
Comment (matsen) [inline src/partition.cpp]: Extra parens

File: src/partition.cpp [hunk 15]
```diff
@@ -87,20 +87,177 @@ Partition::~Partition() {
...
+  root = (pll_utree_t **)malloc(sizeof(pll_utree_t *));
+  SetLabelRoot(minlabel, tree, root);
+  SetLabelRoot(minlabel, tree->back, root);
+  return root[0];
 }
 
+/// @brief Determines order for first ternary step, then runs recursive ordering
+/// for the two non-root subtrees.
+/// @return Completely ordered newick string.
+pll_utree_t *Partition::ToOrderedNewick(pll_utree_t *tree) {
```
Comment (matsen) [inline src/partition.cpp:233]: This also isn't making a Newick string, is it?

File: src/partition.cpp [hunk 16]
```diff
@@ -201,52 +370,144 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
+///@brief Perform every possible NNI move from the current state, and sort them
+/// into good and bad tables.
+void Partition::MakeTables() {
+  // Update and optimize the ML tree, store its logl for comparison, and add it
+  // to the good table.
+  FullBranchOpt(tree_);
+  pll_utree_t *clone = pll_utree_clone(tree_);
+  clone = ToOrderedNewick(clone);
+  double logl = FullTraversalLogLikelihood(clone);
+  good_.insert(ToNewick(clone), logl);
```
Comment (matsen) [inline src/partition.cpp]: Can't figure out how things get added to the bad table!

File: src/partition.cpp [hunk 17]
```diff
@@ -201,52 +370,144 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
+    std::cout << "Log Likelihood for " << item.first << " : " << item.second
+              << std::endl;
+
+  std::cout << "Bad: " << std::endl;
+
+  auto lt1 = bad_.lock_table();
+  for (auto &item : lt1)
+    std::cout << "Log Likelihood for " << item.first << " : " << item.second
+              << std::endl;
+}
```
Comment (matsen) [inline src/partition.cpp:477]: Need space between functions.

File: src/partition.cpp (hunk 16 above)
Comment (matsen) [inline src/partition.cpp]: Oh, I see. This is just the first tree that gets added to the good table. The documentation should make it clear that this function assumes that the current tree is definitely getting added to the good table, e.g. that it's the ML tree.

File: src/partition.cpp [hunk 18]
```diff
@@ -201,52 +369,143 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
+///@param[in] tree
+/// The current edge.
+///@param[in] lambda
+/// The likelihood of ML tree.
+void Partition::NNIComputeEdge(pll_utree_t *tree, double lambda) {
+  TraversalUpdate(tree, 1);
+  // Create a clone of the original tree to perform NNI and reordering on.
+  pll_utree_t *clone = pll_utree_clone(tree);
+  // Set scaler parameter to determine if tree is good/bad.
+  double c = 1.0105;
```
Comment (matsen) [inline src/partition.cpp]: Shouldn't this be an argument? It's eventually going to want to be a command line flag or dynamically set IMHO.

File: src/partition.cpp [hunk 19]
```diff
@@ -201,52 +369,143 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
-  }
-  else{
-  FullTraversalUpdate(tree);
-  OptimizeCurrentBranch(tree);
-  std::cout<<"DONE"<<std::endl;
-
-  TreeBranchLengthsAux(tree->next->back);
-  TreeBranchLengthsAux(tree->next->next->back);
+  if (!tree->next) {
+
```
Comment (matsen) [inline src/partition.cpp]: extra space

File: src/partition.cpp [hunk 20]
```diff
@@ -163,21 +323,18 @@ double Partition::OptimizeCurrentBranch(pll_utree_t* tree) {
...
-      tree->back->scaler_index, tree->pmatrix_index, params_indices_,
-      NULL);
-
-    printf("Branch length: %f log-L: %f Derivative: %f\n", len, opt_logl, d1);
 
+    /// printf("Branch length: %f log-L: %f Derivative: %f D2: %f\n", len,
+    /// opt_logl, d1,d2);
     // If derivative is approximately zero then we've found the maximum.
-    if (fabs(d1) < EPSILON) break;
+    if (fabs(d1) < EPSILON)
```
Comment (matsen) [inline src/partition.cpp:335]: I'd prefer to have every block following an if statement to be wrapped in braces.

File: src/pll-utils.cpp [hunk 21]
```diff
@@ -27,47 +26,29 @@ void fatal(const char *format, ...) {
...
-
-    return (cb_branch_healthy(tree->next->back) &&
-            cb_branch_healthy(tree->next->next->back));
+    if (!tree->next->length || !tree->next->next->length)
+      return 0;
   }
-
-  return true;
-
+  if (tree->length != tree->back->length)
```
Comment (matsen) [inline src/pll-utils.cpp]: Can't we just return a value from the conditional rather than using an if statement?

File: src/pll-utils.cpp [hunk 22]
```diff
@@ -188,77 +172,66 @@ void EquipPartitionWithData(pll_partition_t *partition, pll_utree_t *tree,
...
-/// @todo This should take a RAxML_info file, parse it, and insert those values
-/// here.
+/// @param[in] partition
+/// The model for which to set parameters.
+/// @param[in] path
+/// RAxML info file path for retrieving parameters.
 void SetModelParameters(pll_partition_t *partition, std::string path) {
-  /* initialize the array of base frequencies */
-  /// @todo `Base frequencies` should go here
+  /*Import the info file*/
```
Comment (matsen) [inline src/pll-utils.cpp]: In _our_ code, let's use one convention for comments. Seems like `//` is the way to go.

File: src/test.cpp [hunk 23]
```diff
@@ -1,39 +1,27 @@
 // This tells Catch to provide a main() - only do this in one cpp file.
-#define CATCH_CONFIG_MAIN
+//#define CATCH_CONFIG_MAIN
```
Comment (matsen) [inline src/test.cpp]: Whoa, what happened to the testing framework?

Comment (thayerf) [inline src/test.cpp]: I removed it temporarily and added a main for testing some threads. I can add it back in.

**Comment by matsen** [comment]
@thayerf a lot of great stuff here, but still some work to be done. 

Beyond the little things I've commented on, the big thing is that we need tests. Many more tests. I would like to get to 100% test coverage-- every function has a test associated with it. 

The only reason I'm asking for this is because I really care about this project and want to keep on using the code you are writing!

File: src/partition.cpp [hunk 24]
```diff
@@ -202,21 +203,20 @@ bool Partition::SetLabelRoot(std::string label, pll_utree_t *tree,
     } else
       return false;
   }
-  if (SetLabelRoot(label, tree->next->back, root))
-    return true;
-  if (SetLabelRoot(label, tree->next->next->back, root))
+  if (SetLabelRoot(label, tree->next->back, root) ||
+      SetLabelRoot(label, tree->next->next->back, root))
```
Comment (matsen) [inline src/partition.cpp]: We can just return the value of the conditional, right?

**Comment by thayerf** [comment]
Should I merge this?

**Comment by matsen** [comment]
Are you missing an explicit submodule? 

```
src/partition.hpp:4:43: fatal error: libcuckoo/src/cuckoohash_map.hh: No such file or directory
```

**Comment by matsen** [comment]
I also note that your `Threading` test isn't really a test-- it just prints something out.

**Comment by matsen** [comment]
Compiles! I do get this compiler warning:

```
src/partition.cpp:637:46: warning: explicit by-reference capture of 'good' redundant with by-reference capture default
         pool.push([&, temp, cutoff, lambda, &good, &all, &pool](int id) {
                                              ^
src/partition.cpp:637:53: warning: explicit by-reference capture of 'all' redundant with by-reference capture default
         pool.push([&, temp, cutoff, lambda, &good, &all, &pool](int id) {
                                                     ^
src/partition.cpp:637:59: warning: explicit by-reference capture of 'pool' redundant with by-reference capture default
         pool.push([&, temp, cutoff, lambda, &good, &all, &pool](int id) {
                                                           ^
```

If you can't figure out what's up here, perhaps @bcclaywell can help?

**Comment by bcclaywell** [comment]
```
src/partition.cpp:637:46: warning: explicit by-reference capture of 'good' redundant with by-reference capture default
         pool.push([&, temp, cutoff, lambda, &good, &all, &pool](int id) {
```

This is because the lone ampersand captures everything in the enclosing scope by reference, so saying `[&, &good]` is redundant because `good` is already captured. It's good practice to avoid capturing everything; prefer to capture only the variables the lambda actually needs.

**Pull request closed without merging**